[Feature] aws s3 bucket 연동 구현

hyobin-yang opened the pull request:
# AWS S3 Bucket 연동 구현

## 📋 개요

멀티 클라우드 플랫폼 AgenticCP에 AWS S3 버킷 관리 기능을 헥사고날 아키텍처로 구현했습니다. 테넌트별 자격증명 관리와 포트-어댑터 패턴을 통해 확장 가능하고 테스트 가능한 구조로 설계했습니다.

## 🏗️ 헥사고날 아키텍처 구현

### 1. 포트 인터페이스 (Ports)
- **S3BucketManagementPort**: S3 버킷 생성, 수정, 삭제 기능 정의
- **S3BucketDiscoveryPort**: S3 버킷 조회 및 목록 조회 기능 정의
- **CredentialProviderPort**: 테넌트별 자격증명 해결 인터페이스

### 2. 어댑터 구현 (Adapters)
- **AwsS3BucketManagementAdapter**: AWS S3 API를 통한 버킷 관리 구현
- **AwsS3BucketDiscoveryAdapter**: AWS S3 API를 통한 버킷 조회 구현
- **ThreadLocalCredentialCache**: 멀티 테넌트 자격증명 캐시 관리

### 3. 애플리케이션 서비스 (Use Case)
- **S3BucketUseCaseService**: S3 버킷 관련 비즈니스 로직 처리
- **S3BucketPortRouter**: 프로바이더별 포트 라우팅

### 4. 컨트롤러 (Interface)
- **S3BucketController**: REST API 엔드포인트 제공

## 🔄 주요 구현 흐름

### S3 버킷 생성 흐름
```
Controller → UseCaseService → PortRouter → ManagementAdapter → AWS S3 API
     ↓              ↓              ↓              ↓
  API 요청    비즈니스 로직    포트 라우팅    AWS SDK 호출
```

### S3 버킷 조회 흐름
```
Controller → UseCaseService → PortRouter → DiscoveryAdapter → AWS S3 API
     ↓              ↓              ↓              ↓
  API 요청    비즈니스 로직    포트 라우팅    AWS SDK 호출
```

## ⚙️ 관련 Configuration

### 1. AWS S3 설정 (`AwsS3Config.java`)
```java
@Configuration
public class AwsS3Config {
    @Bean
    public S3Client s3Client() {
        return S3Client.builder()
                .region(Region.of(defaultRegion))
                .credentialsProvider(createThreadLocalCredentialsProvider())
                .build();
    }
}
```

### 2. S3 Capability 설정 (`S3BucketCapabilityConfig.java`)
```java
@Configuration
public class S3BucketCapabilityConfig {
    @PostConstruct
    public void initializeS3BucketCapabilities() {
        CspCapability awsS3Capability = CspCapability.builder()
                .supportsStart(false)
                .supportsStop(false)
                .supportsTerminate(true)
                .supportsTagging(true)
                .supportsListByTag(true)
                .build();
        
        capabilityRegistry.register(ProviderType.AWS, "S3", "BUCKET", awsS3Capability);
    }
}
```

## 🗄️ TenantCloudCredentials 엔티티

### 목적
멀티 테넌트 환경에서 각 테넌트의 클라우드 프로바이더별 자격증명 정보를 안전하게 관리합니다.

### 주요 특징
- **STS AssumeRole 패턴**: IAM Role ARN을 저장하여 보안 강화
- **테넌트별 격리**: 각 테넌트의 자격증명을 독립적으로 관리
- **프로바이더별 지원**: AWS, Azure, GCP 등 다양한 클라우드 프로바이더 지원
- **자격증명 상태 관리**: 활성화/비활성화, 마지막 사용 시간, 오류 추적

### 필드 구조
```java
@Entity
@Table(name = "tenant_cloud_credentials")
public class TenantCloudCredentials extends BaseEntity {
    private String tenantKey;           // 테넌트 키
    private ProviderType providerType;  // 클라우드 프로바이더 타입
    private String accountScope;        // 계정 스코프 (AWS Account ID 등)
    private String roleArn;            // IAM Role ARN
    private String externalId;         // 외부 ID (보안 강화)
    private String sessionName;        // 세션 이름
    private Boolean isActive;          // 활성화 여부
    private LocalDateTime lastUsedAt;  // 마지막 사용 시간
    private String lastErrorMessage;   // 마지막 오류 메시지
}
```

## 🔐 AWS 자격증명 인증 구현

### 1. STS AssumeRole 패턴 🔑
이 패턴은 사용자의 Access Key/Secret Key를 DB에 직접 저장하는 (위험한) 방식 대신, "우리 시스템(AgenticCP)이 사용자의 AWS 계정 내 특정 역할(Role)을 잠시 '빌려 입는(Assume)'" 방식입니다.

#### 상세 구현 흐름:

**1단계: DB 저장**
- AgenticCP의 DB에는 사용자(테넌트)가 등록한 IAM Role의 ARN만 저장
- 예: `arn:aws:iam::[테넌트계정ID]:role/AgenticCP-AccessRole`
- `TenantCloudCredentials` 엔티티에 `roleArn`, `externalId`, `sessionName` 저장

**2단계: resolveCredentials 호출**
- `S3BucketUseCaseService`가 `CredentialProviderPort.resolveCredentials()` 호출
- `AwsCredentialProviderPortAdapter`가 실제 구현 담당

**3단계: AssumeRole 실행**
```java
// AwsCredentialProviderPortAdapter.java
AssumeRoleRequest assumeRoleRequest = AssumeRoleRequest.builder()
    .roleArn(credentials.getRoleArn())
    .roleSessionName(credentials.getDefaultSessionName())
    .durationSeconds(sessionDurationSeconds)
    .externalId(credentials.getExternalId())
    .build();

AssumeRoleResponse assumeRoleResponse = stsClient.assumeRole(assumeRoleRequest);
```

**4단계: 임시 자격증명 획득**
- AWS STS가 1시간 동안 유효한 임시 Access Key, Secret Key, Session Token 발급
- `AwsSessionCredentials` 객체로 래핑

**5단계: ThreadLocal 캐시 저장**
```java
// ThreadLocalCredentialCache에 저장
ThreadLocalCredentialCache.putCredentials(cacheKey, credentials, expirationTime);
```

**6단계: 다운스트림 어댑터 실행**
- `S3BucketUseCaseService` → `router.discovery(...).getBucket(...)`
- `AwsS3BucketDiscoveryAdapter` → `s3Client.headBucket(...)`

**7단계: 자격증명 자동 주입**
- `s3Client`는 `AwsS3Config`에 의해 ThreadLocal 캐시를 먼저 확인
- 저장된 임시 자격증명을 사용하여 AWS API 호출
- **결과적으로 모든 API 호출이 "테넌트의 권한"으로 실행됨**

### 2. ThreadLocal 기반 캐시 관리
- **멀티 테넌트 지원**: 각 요청 스레드별로 독립적인 자격증명 관리
- **자동 만료**: 임시 자격증명의 만료 시간 자동 관리 (5분 여유시간)
- **메모리 효율성**: 요청 완료 시 자동 정리
- **캐시 키 전략**: `tenantKey:providerType:accountScope` 형태

### 3. 보안 강화 요소
- **직접 자격증명 저장 금지**: Access Key/Secret Key를 DB에 저장하지 않음
- **임시 자격증명**: 1시간 후 자동 만료되는 임시 자격증명만 사용
- **External ID**: 추가 보안을 위한 External ID 검증
- **최소 권한 원칙**: 테넌트별로 필요한 최소 권한만 부여
- **감사 추적**: 모든 AssumeRole 요청과 사용 기록 로깅

## 🪣 S3 버킷 관리 및 조회

### 1. 버킷 생성 기능
- **리전별 생성**: 지정된 리전에 버킷 생성
- **객체 소유권 설정**: ACL 비활성화 및 보안 강화
- **객체 잠금 설정**: 버전 관리 및 삭제 보호
- **태그 자동 적용**: 생성 시 태그 자동 설정
- **멱등성 보장**: 이미 존재하는 버킷에 대한 중복 생성 방지

### 2. 버킷 조회 기능
- **Resource Groups Tagging API 활용**: 태그 기반 필터링 지원
- **페이징 지원**: 대용량 버킷 목록 효율적 조회
- **정렬 기능**: 이름, 생성일 등 다양한 기준으로 정렬
- **상세 정보 조회**: 버킷 위치, 태그, 버전 관리 상태 등

### 3. 버킷 삭제 기능
- **안전한 삭제**: 비어있지 않은 버킷 삭제 방지
- **강제 삭제**: 내용물 포함 전체 삭제 옵션
- **버전 관리 지원**: 모든 버전과 삭제 마커 포함 삭제

## 🧪 관련 테스트 코드

### 1. 단위 테스트 (`AwsS3BucketManagementAdapterTest.java`)
- **Mockito 기반**: AWS SDK Mock을 통한 격리된 테스트
- **@Nested 구조**: 기능별 테스트 그룹화
- **예외 상황 테스트**: AWS API 오류 상황 시뮬레이션
- **멱등성 테스트**: 중복 생성 시나리오 검증

### 2. 테스트 커버리지
- **생성 테스트**: 정상 생성, 중복 생성, 권한 오류
- **조회 테스트**: 존재하는 버킷, 존재하지 않는 버킷
- **삭제 테스트**: 일반 삭제, 강제 삭제, 비어있지 않은 버킷
- **업데이트 테스트**: 태그 설정, 버전 관리 설정

## 💰 AWS 연동 테스트 제한사항

### 1. 비용 발생 요소
- **S3 버킷 생성/삭제**: 실제 AWS 리소스 생성으로 인한 비용
- **API 호출**: AWS API 호출에 따른 요청 비용
- **데이터 전송**: 버킷 내용물 조회 시 데이터 전송 비용

### 2. 테스트 제한 사항
- **통합 테스트 제외**: 실제 AWS 환경과의 연동 테스트 제외
- **Mock 기반 테스트**: AWS SDK Mock을 통한 단위 테스트만 수행
- **LocalStack 활용**: 로컬 테스트 환경에서 AWS 서비스 시뮬레이션

### 3. 검증 방법
- **단위 테스트**: 모든 비즈니스 로직과 예외 상황 검증
- **Mock 검증**: AWS SDK 호출 패턴과 파라미터 검증
- **로깅 검증**: 자격증명 해결 및 API 호출 로그 검증


## 📝 향후 개선 사항

1. **통합 테스트 환경**: LocalStack을 활용한 실제 AWS API 테스트
2. **자격증명 최적화**: 캐시 전략 및 만료 시간 관리 개선
3. **에러 처리 강화**: 더 세밀한 AWS 오류 코드 매핑
4. **모니터링 추가**: 자격증명 사용량 및 성능 모니터링

## 🚨 CredentialProviderPort 관련

### CredentialProviderPort 파라미터 변경: tenantId → tenantKey
`CredentialProviderPort` 인터페이스에서 `tenantId` 대신 `tenantKey`를 사용하는 이유:

- **비즈니스 식별자 사용**: `tenantId`는 내부 DB의 Primary Key이지만, `tenantKey`는 비즈니스적으로 의미있는 고유 식별자
- **API 일관성**: 외부 API와의 통신에서 더 직관적이고 안전한 식별자 사용
- **보안 강화**: 내부 DB ID 노출 방지 및 비즈니스 로직과의 명확한 분리
- **확장성**: 향후 멀티 테넌트 환경에서 더 유연한 식별자 관리 가능

## 🚨 Cache Manager @bean 관련

- HealthCheckCacheConfig, RedisCacheConfig에서 빈 등록 시 충돌 발생 방지 위해 빈 name을 명시

hyobin-yang committed:
feat: S3 버킷 도메인 엔티티 및 모델 추가

- S3BucketIdentity, CreateS3BucketRequest, UpdateS3BucketRequest 모델 추가
- S3BucketQuery 쿼리 모델 구현

hyobin-yang committed:
feat: S3 버킷 관리 포트 인터페이스 정의

- S3BucketDiscoveryPort: 버킷 탐색 기능
- S3BucketManagementPort: 버킷 CRUD 작업

hyobin-yang committed:
feat: S3 버킷 관리 어댑터 및 매퍼 구현

hyobin-yang committed:
feat: s3 bucket 관련 config 클래스 추가

hyobin-yang committed:
feat: s3 bucket CRUD 컨트롤러 및 서비스 구현

hyobin-yang committed:
chore: s3 bucket 관련 설정 추가

hyobin-yang committed:
test: s3 bucket CRUD 관련 서비스 테스트

hyobin-yang committed:
feature: aws,s3 로직 관련 로직 에러 코드 추가

hyobin-yang committed:
chore: cacheManager의 빈 이름 명시하여 충돌 방지

hyobin-yang committed:
refactor: cursor rule 적용하여 리팩토링 및 파일 경로 변경

hyobin-yang committed:
refactor: command 계층 이용하여 dto 의존성 끊기

hyobin-yang committed:
chore: cacheManager의 빈 이름 명시하여 충돌 방지

hyobin-yang committed:
feat: 빈 주입 에러 방지 위한 포트 어댑터 틀 구현

hyobin-yang committed:
feat:  AWS 자격증명 기능 구현 및 적용

hyobin-yang committed:
chore: AWS 자격증명 관련 의존성 추가

hyobin-yang committed:
refactor: AwsS3BucketMapper 구조 개선

hyobin-yang committed:
test: s3 bucket 연동 관련 단위테스트 추가

hyobin-yang commented:
# S3 Bucket Management API 스펙

## 개요

S3 버킷 관리를 위한 REST API 컨트롤러입니다. AWS S3 버킷의 생성, 조회, 수정, 삭제 등의 기능을 제공합니다.

### 기본 정보
- **Base URL**: `/api/v1/cloud/s3/buckets`
- **Tag**: S3 Bucket Management
- **설명**: S3 버킷 관리 API
- **Controller**: `S3BucketController.java`

---

## API 목록

### 1. S3 버킷 생성

**POST** `/api/v1/cloud/s3/buckets`

새로운 S3 버킷을 생성합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| request | CreateS3BucketRequest | Body | ✅ | S3 버킷 생성 요청 데이터 | - |

#### 권한 요구사항
- `S3_BUCKET_CREATE` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 201 | 버킷 생성 성공 |
| 400 | 잘못된 요청 데이터 |
| 409 | 버킷 이름 중복 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `CREATE_BUCKET`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `HIGH`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ✅

---

### 2. S3 버킷 업데이트

**PUT** `/api/v1/cloud/s3/buckets/{bucketName}`

기존 S3 버킷의 설정을 업데이트합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| bucketName | String | Path | ✅ | 버킷 이름 | my-bucket |
| request | UpdateS3BucketRequest | Body | ✅ | S3 버킷 업데이트 요청 데이터 | - |

#### 권한 요구사항
- `S3_BUCKET_UPDATE` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 200 | 버킷 업데이트 성공 |
| 400 | 잘못된 요청 데이터 |
| 404 | 버킷을 찾을 수 없음 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `UPDATE_BUCKET`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `HIGH`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ✅

---

### 3. S3 버킷 목록 조회

**GET** `/api/v1/cloud/s3/buckets`

지정된 클라우드 프로바이더의 S3 버킷 목록을 조회합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 기본값 | 설명 | 예시 |
|---------|------|------|------|--------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | - | 클라우드 프로바이더 타입 | AWS |
| page | int | Query | ❌ | 0 | 페이지 번호 (0부터 시작) | 0 |
| size | int | Query | ❌ | 10 | 페이지 크기 | 10 |

#### 권한 요구사항
- `S3_BUCKET_READ` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 200 | 버킷 목록 조회 성공 |
| 400 | 잘못된 요청 파라미터 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `LIST_BUCKETS`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `LOW`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ❌

---

### 4. S3 버킷 상세 조회

**GET** `/api/v1/cloud/s3/buckets/{bucketName}`

지정된 S3 버킷의 상세 정보를 조회합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| bucketName | String | Path | ✅ | 버킷 이름 | my-bucket |

#### 권한 요구사항
- `S3_BUCKET_READ` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 200 | 버킷 조회 성공 |
| 404 | 버킷을 찾을 수 없음 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `GET_BUCKET`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `LOW`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ❌

---

### 5. S3 버킷 존재 확인

**GET** `/api/v1/cloud/s3/buckets/{bucketName}/exists`

지정된 S3 버킷의 존재 여부를 확인합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| bucketName | String | Path | ✅ | 버킷 이름 | my-bucket |

#### 권한 요구사항
- `S3_BUCKET_READ` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 200 | 존재 확인 성공 (Boolean 값 반환) |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `CHECK_BUCKET_EXISTS`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `LOW`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ❌

---

### 6. S3 버킷 삭제

**DELETE** `/api/v1/cloud/s3/buckets/{bucketName}`

지정된 S3 버킷을 삭제합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| bucketName | String | Path | ✅ | 버킷 이름 | my-bucket |

#### 권한 요구사항
- `S3_BUCKET_DELETE` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 204 | 버킷 삭제 성공 |
| 404 | 버킷을 찾을 수 없음 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `DELETE_BUCKET`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `CRITICAL`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ❌

---

### 7. S3 버킷 강제 삭제

**DELETE** `/api/v1/cloud/s3/buckets/{bucketName}/force`

지정된 S3 버킷을 내용물과 함께 강제 삭제합니다.

#### 요청 파라미터

| 파라미터 | 타입 | 위치 | 필수 | 설명 | 예시 |
|---------|------|------|------|------|------|
| provider | CloudProvider.ProviderType | Query | ✅ | 클라우드 프로바이더 타입 | AWS |
| bucketName | String | Path | ✅ | 버킷 이름 | my-bucket |

#### 권한 요구사항
- `S3_BUCKET_DELETE` 권한 또는 `ADMIN` 역할

#### 응답 코드

| 코드 | 설명 |
|------|------|
| 204 | 버킷 강제 삭제 성공 |
| 404 | 버킷을 찾을 수 없음 |
| 500 | 서버 내부 오류 |

#### 감사 로그
- **액션**: `FORCE_DELETE_BUCKET`
- **리소스 타입**: `S3_BUCKET`
- **심각도**: `CRITICAL`
- **요청 데이터 포함**: ✅
- **응답 데이터 포함**: ❌

---

## 공통 특징

### 응답 형식
모든 API는 `ApiResponse<T>` 형식으로 응답합니다.

```json
{
  "success": true,
  "message": "성공 메시지",
  "data": { /* 응답 데이터 */ },
  "timestamp": "2024-01-01T00:00:00Z"
}
```

### 로깅
- 모든 API는 진입 시와 성공 시에 로그를 남깁니다.
- 민감한 정보는 마스킹 처리됩니다.

### 보안
- Spring Security의 `@PreAuthorize`를 사용하여 권한 기반 접근 제어를 구현했습니다.
- 각 API별로 필요한 권한이 정의되어 있습니다.

### 감사 로깅
- 모든 API는 `@AuditRequired` 어노테이션을 통해 감사 로그를 자동으로 기록합니다.
- 액션, 리소스 타입, 심각도, 데이터 포함 여부가 설정되어 있습니다.

### 에러 처리
- 비즈니스 예외는 `BusinessException` 계층을 사용합니다.
- `GlobalExceptionHandler`에서 자동으로 `ApiResponse` 형식으로 변환됩니다.

---

## 사용 예시

### 버킷 생성
```bash
curl -X POST "http://localhost:8080/api/v1/cloud/s3/buckets?provider=AWS" \
  -H "Content-Type: application/json" \
  -H "Authorization: Bearer <token>" \
  -d '{
    "bucketName": "my-test-bucket",
    "region": "us-east-1"
  }'
```

### 버킷 목록 조회
```bash
curl -X GET "http://localhost:8080/api/v1/cloud/s3/buckets?provider=AWS&page=0&size=10" \
  -H "Authorization: Bearer <token>"
```

### 버킷 존재 확인
```bash
curl -X GET "http://localhost:8080/api/v1/cloud/s3/buckets/my-bucket/exists?provider=AWS" \
  -H "Authorization: Bearer <token>"
```

---

## 주의사항

1. **권한 확인**: 각 API 호출 전에 적절한 권한이 있는지 확인해야 합니다.
2. **감사 로그**: 모든 작업이 감사 로그에 기록되므로 신중하게 사용해야 합니다.
3. **강제 삭제**: `forceDeleteBucket` API는 버킷의 모든 내용물을 삭제하므로 주의가 필요합니다.
4. **프로바이더**: 모든 API에서 `provider` 파라미터는 필수입니다.
5. **버킷 이름**: S3 버킷 이름은 전 세계적으로 고유해야 합니다.

moonhyeonmin commented on src/main/java/com/agenticcp/core/domain/cloud/adapter/outbound/common/TracingPortAdapter.java:
AuditEventPortAdapter, CredentialProviderPortAdapter, TracingPortAdapter 공통으로 쓰이는 어댑터인 것 같은데, 효빈님 코드랑 동일하게 구현하면 될까용???

moonhyeonmin reviewed (commented):
코드가 너무 깔끔합니다 한 수 배웁니다

KwonSunJae committed:
Merge remote-tracking branch 'origin/develop' into feature/129

KwonSunJae committed:
fix: Update RedisCacheConfig to use @Primary annotation directly

- Changed the annotation for the cacheManager bean from the fully qualified name to the simpler @Primary import.
- This improves readability and maintains consistency in the configuration code.
